app/javascript/mastodon/features/video/index.tsx

```diff
@@ -0,0 +1,1022 @@
+  skipForward: { id: 'video.skip_forward', defaultMessage: 'Skip forward' },
+  skipBackward: { id: 'video.skip_backward', defaultMessage: 'Skip backward' },
+});
+
+const DOUBLE_CLICK_THRESHOLD = 250;
+
+export const formatTime = (secondsNum: number) => {
+  const hours = Math.floor(secondsNum / 3600);
+  const minutes = Math.floor((secondsNum - hours * 3600) / 60);
+  const seconds = secondsNum - hours * 3600 - minutes * 60;
+
+  const formattedHours = `${hours < 10 ? '0' : ''}${hours}`;
+  const formattedMinutes = `${minutes < 10 ? '0' : ''}${minutes}`;
+  const formattedSeconds = `${seconds < 10 ? '0' : ''}${seconds}`;
+
+  return (
+    (formattedHours === '00' ? '' : `${formattedHours}:`) +
+    `${formattedMinutes}:${formattedSeconds}`
+  );
+};
+
+export const findElementPosition = (el: HTMLElement) => {
+  const box = el.getBoundingClientRect();
+  const docEl = document.documentElement;
+  const body = document.body;
+
+  const clientLeft = docEl.clientLeft || body.clientLeft || 0;
+  const scrollLeft = window.scrollX || body.scrollLeft;
+  const left = box.left + scrollLeft - clientLeft;
+
+  const clientTop = docEl.clientTop || body.clientTop || 0;
+  const scrollTop = window.scrollY || body.scrollTop;
+  const top = box.top + scrollTop - clientTop;
+
+  return {
+    left: Math.round(left),
+    top: Math.round(top),
+  };
+};
+
+export const getPointerPosition = (
+  el: HTMLElement | null,
+  event: MouseEvent,
+) => {
+  if (!el) {
+    return {
+      y: 0,
+      x: 0,
+    };
+  }
+
+  const box = findElementPosition(el);
+  const boxW = el.offsetWidth;
+  const boxH = el.offsetHeight;
+  const boxY = box.top;
+  const boxX = box.left;
+
+  const { pageY, pageX } = event;
+
+  return {
+    y: Math.max(0, Math.min(1, (pageY - boxY) / boxH)),
+    x: Math.max(0, Math.min(1, (pageX - boxX) / boxW)),
+  };
+};
+
+export const fileNameFromURL = (str: string) => {
+  const url = new URL(str);
+  const pathname = url.pathname;
+  const index = pathname.lastIndexOf('/');
+
+  return pathname.slice(index + 1);
+};
+
+const frameRateAsNumber = (frameRate: string): number => {
+  if (frameRate.includes('/')) {
+    return frameRate
+      .split('/')
+      .map((c) => parseInt(c))
+      .reduce((p, c) => p / c);
+  }
+
+  return parseInt(frameRate);
+};
+
+const persistVolume = (volume: number, muted: boolean) => {
+  playerSettings.set('volume', volume);
+  playerSettings.set('muted', muted);
+};
+
+const restoreVolume = (video: HTMLVideoElement) => {
+  const volume = (playerSettings.get('volume') as number | undefined) ?? 0.5;
+  const muted = (playerSettings.get('muted') as boolean | undefined) ?? false;
+
+  video.volume = volume;
+  video.muted = muted;
+};
+
+let hotkeyEventId = 0;
+
+const registerHotkeyEvent = (
+  setHotkeyEvents: React.Dispatch<React.SetStateAction<HotkeyEvent[]>>,
+  event: Omit<HotkeyEvent, 'key'>,
+) => {
+  setHotkeyEvents(() => [{ key: hotkeyEventId++, ...event }]);
+};
```

Comment:
nit: perhaps this should be moved inside Video so the code doesn't need to pass `setHotkeyEvents` every time?

Reply:
I still get a bit confused about `useCallback` depending on another `useCallback` so wanted to avoid that by extracting the helper function outside the component. The most important bit here (and I think I should've left a comment maybe) is that it's important the `event` literal is instantiated *outside* the `setHotkeyEvents` callback because it runs at a later time and so the `videoRef.current.paused` value will already be different. By passing the `event` literal to this function instead, I keep it all on one line where it's called, and the `event` value is static.